File: th/codeforces/weighted_graph/kruskal.py
import sys
input = sys.stdin.readline
def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    n, m = int(data[idx]), int(data[idx + 1])
    idx += 2
    edges = []
    for _ in range(m):
        u, v, w = int(data[idx]), int(data[idx + 1]), int(data[idx + 2])
        idx += 3
        if u != v:  # bỏ self-loop
            edges.append((w, u, v))
    # Sắp xếp theo trọng số
    edges.sort()
    # DSU (Disjoint Set Union) với path compression + union by rank
    parent = list(range(n + 1))
    rank   = [0] * (n + 1)
    def find(x):
        # Iterative path compression
        root = x
        while parent[root] != root:
            root = parent[root]
        while parent[x] != root:
            parent[x], x = root, parent[x]
        return root
    def union(x, y):
        rx, ry = find(x), find(y)
        if rx == ry:
            return False
        if rank[rx] < rank[ry]:
            rx, ry = ry, rx
        parent[ry] = rx
        if rank[rx] == rank[ry]:
            rank[rx] += 1
        return True
    total_weight = 0
    edge_count   = 0
    for w, u, v in edges:
        if union(u, v):
            total_weight += w
            edge_count   += 1
            if edge_count == n - 1:
                break
    if edge_count < n - 1:
        print(-1)
    else:
        print(total_weight)
main()

File: th/codeforces/weighted_graph/kruskal.cpp
#include <bits/stdc++.h>
using namespace std;

// ── Fast I/O (fread-based) ────────────────────────────────────────────────────
namespace fastio {
    static char buf[1 << 22]; // 4 MB buffer
    static int  pos, len;

    void init() { len = fread(buf, 1, sizeof(buf), stdin); }

    inline long long readLL() {
        // skip non-numeric chars (space, newline, etc.)
        while (pos < len && buf[pos] != '-' && (buf[pos] < '0' || buf[pos] > '9'))
            pos++;
        if (pos >= len) return 0;
        bool neg = (buf[pos] == '-');
        if (neg) pos++;
        long long x = 0;
        while (pos < len && buf[pos] >= '0' && buf[pos] <= '9')
            x = x * 10 + (buf[pos++] - '0');
        return neg ? -x : x;
    }

    inline int readInt() { return (int)readLL(); }
}

// ── DSU ───────────────────────────────────────────────────────────────────────
static int parent[200005];
static int rnk[200005];

inline int find(int x) {
    while (parent[x] != x) {
        parent[x] = parent[parent[x]]; // path halving
        x = parent[x];
    }
    return x;
}

inline bool unite(int x, int y) {
    int px = find(x), py = find(y);
    if (px == py) return false;
    if (rnk[px] < rnk[py]) swap(px, py);
    parent[py] = px;
    if (rnk[px] == rnk[py]) rnk[px]++;
    return true;
}

// ── Edge (dùng long long cho weight) ─────────────────────────────────────────
struct Edge {
    long long w;
    int u, v;
    bool operator<(const Edge& o) const { return w < o.w; }
};

// ── Main ──────────────────────────────────────────────────────────────────────
int main() {
    fastio::init();

    int n = fastio::readInt();
    int m = fastio::readInt();

    vector<Edge> edges(m);
    for (auto& e : edges) {
        e.u = fastio::readInt();
        e.v = fastio::readInt();
        e.w = fastio::readLL();   // long long để tránh overflow
    }

    sort(edges.begin(), edges.end()); // O(m log m)

    iota(parent + 1, parent + n + 1, 1); // parent[i] = i

    long long total = 0;
    int used = 0;

    for (auto& e : edges) {
        if (unite(e.u, e.v)) {
            total += e.w;
            if (++used == n - 1) break;
        }
    }

    printf("%lld\n", used < n - 1 ? -1LL : total);
    return 0;
}
